### PR TITLE
Additional cleanup and safety checks for Android

### PR DESCRIPTION
**Pull request by blair1618**
- Checks the validity of every value used as a native pointer and throws Exceptions for bad values
 - Calls remove() on every MapData associated with a MapController in dispose()

### CODE REVIEW

File: android/tangram/src/com/mapzen/tangram/MapController.java [hunk 1]
```diff
@@ -184,6 +184,9 @@ void dispose() {
         queueEvent(new Runnable() {
             @Override
             public void run() {
+                for (Map.Entry<String, MapData> entry : clientDataSources.entrySet()) {
```
Comment (tallytalwar) [inline android/tangram/src/com/mapzen/tangram/MapController.java]: Can you also do a `checkPointer(mapPointer)` when map is init. Line: 179.

P.S. Github did not allow me to comment of that specific line!!

File: android/tangram/src/com/mapzen/tangram/MapData.java [hunk 2]
```diff
@@ -49,8 +49,9 @@ public String name() {
     }
 
     /**
-     * Remove this {@code MapData} from the map it is currently associated with. After
-     * {@code remove} is called, this object can no longer apply changes to the map.
+     * Remove this {@code MapData} from the map it is currently associated with. Using this object
+     * after {@code remove} is called will cause a crash. {@code remove} is called on every
```
Comment (tallytalwar) [inline android/tangram/src/com/mapzen/tangram/MapData.java]: will cause an `exception` instead of a `crash`.

**Comment by tallytalwar** [comment]
Good to go, once couple of commits are addressed.

**Comment by blair1618** [comment]
Thanks, updated!

**Pull request merged**